File: src/java/Boot/src/com/core/api/module_server/Scales.java
package com.core.api.module_server;

import com.zhekasmirnov.innercore.api.NativeAPI;

import org.mozilla.javascript.ScriptableObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;

public class Scales {
    public static interface IScaleDescription {
        String getName();
        String getFull();
        String getHelf();
        String getEmpty();
        boolean isLeft();
        boolean isReset();
        boolean isDisplay();
    }

    public static class ScalePlayer {
        Scales pointer;
        int value;

        public ScalePlayer(Scales pointer){
            this.pointer = pointer;
        }

        public void setValue(int value){
            this.value = value;
        }

        public int getValue(){
            return value;
        }

        public Scales getType(){
            return pointer;
        }

        public void reset(){
            value = 20;
        }
    }

    private static final HashMap<String, HashMap<String, ScalePlayer>> players = new HashMap<>();
    private static final HashMap<String, Scales> scales = new HashMap<>();

    public static String[] getPlayers(){
        final ArrayList<String> keys = new ArrayList<>();
        final Iterator<String> it = players.keySet().iterator();

        while (it.hasNext()) {
            keys.add(it.next());
        }

        return keys.toArray(new String[0]);
    }

    public static String[] getScales(){
        final ArrayList<String> keys = new ArrayList<>();
        final Iterator<String> it = scales.keySet().iterator();

        while (it.hasNext()) {
            keys.add(it.next());
        }

        return keys.toArray(new String[0]);
    }

    public static Scales getScaleByName(String name){
        return scales.get(name);
    }

    public static ScalePlayer getScaleByPlayer(long player, String name){
        return getScaleByPlayer(NativeAPI.getNameTag(player), name);
        //return new ScalePlayer(getScalePlayer(NativeAPI.getNameTag(player), name));
    }

    public static ScalePlayer getScaleByPlayer(String player, String name){
        HashMap<String, ScalePlayer> datas = players.get(player);
        if(datas == null){
            datas = new HashMap<>();
            players.put(player, datas);
        }

        ScalePlayer data = datas.get(name);
        if(data == null){
            data = new ScalePlayer(getScaleByName(name));
            datas.put(name, data);
        }

        return data;
    }

    public static Scales register(ScriptableObject description){
        return new Scales(description);
    }

    public static void registerScale(String name, boolean left, boolean reset, boolean display, Scales self){
        self.name = name;
        self.isLeft = left;
        self.isReset = reset;
        self.isDisplay = display;
        scales.put(name, self);
    }

    private String name;
    private boolean isLeft, isReset, isDisplay;

    public Scales(IScaleDescription description){
        registerScale(
            description.getName(),
            description.isLeft(),
            description.isReset(),
            description.isDisplay(),
            this
        );
    }

    public Scales(ScriptableObject object){
        registerScale(
            (String) object.get("name"),
            (boolean) object.get("isLeft"),
            (boolean) object.get("isReset"),
            object.has("isDisplay", object) ? (boolean) object.get("isDisplay") : true,
            this
        );
    }

    public String getName(){
        return name;
    }

    public String getFull(){
        return "";
    }

    public void setFull(String name){
    }

    public String getHelf(){
        return "";
    }

    public void setHelf(String name){
    }

    public String getEmpty(){
        return "";
    }

    public void setEmpty(String name){
    }

    public boolean isLeft(){
        return isLeft;
    }

    public boolean isReset(){
        return isReset;
    }

    public boolean isDisplay(){
        return isDisplay;
    }

    public void setDisplay(boolean v){
       isDisplay = v;
    }

    public IScaleDescription getScaleDescription(){
        Scales _this = this;
        return new IScaleDescription() {
            @Override
            public String getName() {
                return _this.getName();
            }
            @Override
            public String getFull() {
                return _this.getFull();
            }
            @Override
            public String getHelf() {
                return _this.getHelf();
            }
            @Override
            public String getEmpty() {
                return _this.getEmpty();
            }
            @Override
            public boolean isLeft() {
                return _this.isLeft();
            }
            @Override
            public boolean isReset() {
                return _this.isReset();
            }
            @Override
            public boolean isDisplay() {
                return _this.isDisplay();
            }
        };
    }

    public ScriptableObject getScriptable(){
        ScriptableObject object = new ScriptableObject(){
            @Override
            public String getClassName() {
                return "null";
            }
        };
        object.put("name", object, getName());
        object.put("full", object, getFull());
        object.put("helf", object, getHelf());
        object.put("empty", object, getEmpty());
        object.put("isLeft", object, isLeft());
        object.put("isReset", object, isReset());
        return object;
    }
}

File: src/dev/api/Scales.ts
type JavaPlayerScale = {
	setValue(value: number): void;
	getValue(): number;
	reset(): void;
	getType(): JavaScale;
}

type JavaScale = {
	setDisplay(v: boolean);
	isDisplay(): boolean;
	getScaleDescription(): ScaleDescription;
	isReset(): boolean;
}

interface ScaleDescription {
	name: string,
	isLeft: boolean,
	isReset: boolean,
	isDisplay?: boolean
}

type JavaScales = {
	getPlayers(): string[];
	getScales(): string[];
	getScaleByName(name: string): JavaScale;
	getScaleByPlayer(player: number | string, name: string): JavaPlayerScale;
	register(description: ScaleDescription): JavaScale;

	setServerMode(value: boolean);
}

type PacketScalesSetValue = {
	player: string,
	name: string,
	value: number
}

type PacketScalesSetServerMode = {
	mode: boolean
}

type SaveScope = {
	players: {[player: string]: {[scale_name: string]: number}}
}

if(Game.isDedicatedServer())
	var ScalesJava: JavaScales = WRAP_JAVA("com.core.api.module_server.Scales")
else
	var ScalesJava: JavaScales = WRAP_JAVA("com.core.api.module.Scales");


Network.addClientPacket("scales.setValue", function(data: PacketScalesSetValue){
	ScalesJava.getScaleByPlayer(data.player, data.name).setValue(data.value);
}); 

Network.addClientPacket("scales.setMode", function(data: PacketScalesSetServerMode){
	ScalesJava.setServerMode(data.mode);
}); 


Callback.addCallback("EntityDeath", (entity) => {
	let players: any = Network.getConnectedPlayers();
	if(players.indexOf(entity) != -1){
		let scales = ScalesJava.getScales();
		for(let i in scales){
			let name = scales[i];
			if(ScalesJava.getScaleByName(name).isReset())
				ScalesJava.getScaleByPlayer(entity, name).reset();
		}
	}
});

Callback.addCallback("ServerPlayerLoaded", (player) => {
	let scales = ScalesJava.getScales();
	let client = Network.getClientForPlayer(player);

	if(client){
		for(let i in scales)
			Scales.getScaleByPlayer(player, scales[i]).send();
		let packet: PacketScalesSetServerMode = {mode: Game.isDedicatedServer()};
		client.send("scales.setMode", packet)
	};
	
});

class PlayerScale {
	private java_class: JavaPlayerScale;
	private player: number;
	private name: string;

	constructor(java_class: JavaPlayerScale, player: number, name: string){
		this.java_class = java_class;
		this.player = player;
		this.name = name;
	}

	public getValue(): number {
		return this.java_class.getValue();
	}

	public send(value: number = this.getValue(), client: NetworkClient = Network.getClientForPlayer(this.player)): void {
		let packet: PacketScalesSetValue = {//конвертация типов java в js 
			player: Entity.getNameTag(this.player),
			name: String(this.name),
			value: Number(value)
		}
		client && client.send("scales.setValue", packet);
	}

	public setValue(value: number): void {
		this.send(value);
		this.java_class.setValue(value);
	}

	public addValue(v: number): void {
		this.setValue(this.getValue() + v);
	}

	public reset(): void {
		this.java_class.reset();
	}

	public getType(): JavaScale {
		return this.java_class.getType();
	}
}

class Scale {
	private java_class: JavaScale;

	constructor(java_class: JavaScale){
		this.java_class = java_class;
	}

	public setDisplay(v: boolean): Scale {
		this.java_class.setDisplay(v);
		return this;
	}

	public isDisplay(): boolean {
		return this.java_class.isDisplay();
	}
}

namespace Scales {
	export function register(obj: ScaleDescription): Scale {
		return new Scale(ScalesJava.register(obj));
	} 

	export function getScaleByPlayer(player: number, name: string){
		return new PlayerScale(ScalesJava.getScaleByPlayer(player, name), player, name);
	}

	export function isCreative(player: number, name: string): boolean {
		return new PlayerActor(player).getGameMode() == 0;
	}
}

Saver.addSavesScope("lib.scales", (scope: SaveScope) => {
    let players = Object.keys(scope.players || {});
		
	for(let i in players){
		let player = players[i];
		let scales = Object.keys(scope.players[player]);

		for(let j in scales){
			let scale = scales[j];
			ScalesJava.getScaleByPlayer(player, scale).setValue(scope.players[player][scale]);
		}
	}
}, () => {
    let obj = {};
	let players = ScalesJava.getPlayers();

	for(let i in players){
		let player = players[i];
		let scales = ScalesJava.getScales();
		let data_save = obj[player] = {};

		for(let j in scales){
			let scale = scales[j];
			data_save[scale] = ScalesJava.getScaleByPlayer(player, scale).getValue();
		}
	}

	alert(JSON.stringify(obj))
    return {
        players: obj,
    }
});